android/src/main/java/com/vydia/RNUploader/UploadWorker.kt

```diff
@@ -0,0 +1,201 @@
+        onUpload { bytesSentTotal, _ ->
+          // throttle progress report
+          val now = System.currentTimeMillis()
+          if (now - lastProgressReport >= PROGRESS_INTERVAL) {
+            lastProgressReport = now
+            handleProgress(bytesSentTotal, size)
+          }
+        }
+      }
+
+      return@withContext handleSuccess(response, size)
+    } catch (error: Throwable) {
+      return@withContext handleError(error, retry = true, retries)
+    } finally {
+      // stop waiting
+      semaphore.release()
+    }
+  }
+
+  private suspend fun handleProgress(bytesSentTotal: Long, fileSize: Long) {
+    UploadProgress.set(context, upload.id, bytesSentTotal, fileSize)
+    EventReporter.progress(upload.id, bytesSentTotal, fileSize)
+    setForeground(getForegroundInfo())
+  }
+
+  private fun handleSuccess(response: HttpResponse, fileSize: Long): Result {
+    UploadRetry.clear(context, upload.id)
+    UploadProgress.set(context, upload.id, fileSize, fileSize)
+    UploadProgress.scheduleClearing(context)
+    EventReporter.success(upload.id, response)
+    return Result.success()
+  }
+
+  private fun handleError(error: Throwable, retry: Boolean, retries: Int = 0): Result {
+    // Cancelled by user or new worker with same ID
+    // Worker won't rerun, perform teardown
+    if (isStopped) {
+      UploadRetry.clear(context, upload.id)
+      UploadProgress.remove(context, upload.id)
+      UploadProgress.scheduleClearing(context)
+      EventReporter.cancelled(upload.id)
+      throw error
+    }
+
+    // Auto-cancelled, likely due to unmet constraints
+    // Clearing state not needed, worker will restart
+    if (error is CancellationException) throw error
+
+    // retry if possible
+    if (retry && retries < upload.maxRetries) {
+      UploadRetry.set(context, upload.id, retries + 1)
+      return Result.retry()
+    }
+
+    // no more retrying
+    UploadRetry.clear(context, upload.id)
+    UploadProgress.remove(context, upload.id)
+    UploadProgress.scheduleClearing(context)
+    EventReporter.error(upload.id, error)
+    return Result.failure()
+  }
+
+  // builds the notification required to enable Foreground mode
+  override suspend fun getForegroundInfo(): ForegroundInfo {
+    // All workers share the same notification that shows the total progress
+    val id = upload.notificationId.hashCode()
+    val title = upload.notificationTitle
+    val channel = upload.notificationChannel
+    val progress = UploadProgress.total(context)
+
+    // Custom layout for progress notification.
+    // The default hides the % text. This one shows it on the right,
+    // like most examples in various docs.
+    val content = RemoteViews(context.packageName, R.layout.notification)
+    content.setTextViewText(R.id.notification_title, title)
+    content.setTextViewText(R.id.notification_progress, "$progress%")
+    content.setProgressBar(R.id.notification_progress_bar, 100, progress, false)
```

Comment:
here's a screenshot of the custom view
![IMG_1809](https://user-images.githubusercontent.com/2142140/235270618-fe4211a4-435c-4a2c-8d62-fd00967f2854.jpg)